### ISSUE TITLE
Performance issue in lazy grid

### ISSUE THREAD

**Issue by PauliusVal** [coil-kt/coil#1866]
**Description**
Compose performance issue in super simple app which just [displays network images list using Coil](https://github.com/PauliusVal/ComposeGridPerformance/blob/main/app/src/main/java/net/composegridperformance/feature/images/ImagesGridScreen.kt).

We see noticeable jank while scrolling even though:
- app runs in R8 release mode
- contains baseline profile for scrolling screen
- stable Compose parameters used
- unique and stable item keys applied for grid items

Jank is visible by human eye and also in profiler:
https://github.com/PauliusVal/ComposeGridPerformance/blob/main/images/images_grid_janky_frames_1.png?raw=true

Full app code and more screenshots: https://github.com/PauliusVal/ComposeGridPerformance/tree/main

**Version**
Coil version: `io.coil-kt:coil-compose:2.4.0`

**Comment by colinrtwhite** [comment]
Thanks for the repro project! Which device did you run the tests on?

**Comment by PauliusVal** [comment]
Samsung S21 Ultra, Android 13

**Comment by PauliusVal** [comment]
Not sure if that helps, but updated the app with identical grid just with Glide image loading. It has noticeably fewer janky frames in profiler and looks/feels much smoother.

**Comment by Mkohm** [comment]
Same problem here, seems to use a lot of memory! Up to 2gigs of memory in my app when scrolling in lazy column with images.

**Comment by lkhore** [comment]
Same issue janky frames with grid

**Comment by colinrtwhite** [comment]
@Mkohm That sounds like it could be a separate memory leak. Please file an issue (ideally with a way to reproduce)!

@lkhore Did you test in release mode? Compose is often much slower in debug mode.

**Comment by Kuki93** [comment]
Is there any progress on this issue

**Comment by matejdro** [comment]
I've added benchmark and tracing here to make diagnosing this easier: https://github.com/matejdro/ComposeGridPerformance/commit/beb902c3ad6d1348c169fc4565524fcaf0da5c5


Here is an example trace file: [ScrollBenchmark_scrollImageList_iter003_2023-12-21-08-27-41.perfetto-trace.zip](https://github.com/coil-kt/coil/files/13739015/ScrollBenchmark_scrollImageList_iter003_2023-12-21-08-27-41.perfetto-trace.zip)

Not sure why the trace is so weird (it shows one big compose event spanning multiple seconds), but you can still drill down and it appears that `rememberAsyncPainter` is taking considerable time:

![image](https://github.com/coil-kt/coil/assets/507922/3cda3a30-71b0-44b3-a361-181513d777ea)

**Comment by hellosagar** [comment]
Im noticing a heavy jank in scrolling with LazyGrid as well

**Comment by musooff** [comment]
I am simple loading local images and showing with coil. There is a noticeable lagging (skipping lots of frames) especially in older devices. 
During first showing of the LazyVerticalGrid it is almost unusable. 
Below is the video to show the lag with HWUI rendering. 
And yes, it is in release mode. In debug you can't even scroll. 

https://github.com/coil-kt/coil/assets/22423682/f3b42d97-6d29-438e-85c6-5cd009a028bd

**Comment by mainrs** [comment]
Hope I can somehow help with this. I am using `LazyColumn` and each item has an `AsyncImage` ("episode item"). Is there anything I can provide: debug logs, source code, ....?

**Comment by x3x0z** [comment]
I have the same error, I'm using a LazyVerticalGrid with +8k items (I'm also using pagination), after fast scrolling through more than 3k items, the application gets super slow, even the android emulator stops working.

**Comment by Cj-Rodriguez101** [comment]
> Hope I can somehow help with this. I am using `LazyColumn` and each item has an `AsyncImage` ("episode item"). Is there anything I can provide: debug logs, source code, ....?

Does it still happen on 2.6.0?

**Comment by mikelpr** [comment]
@Cj-Rodriguez101 happens to me with a `LazyVerticalGrid` of `AsyncImages` on 2.6.0

**Comment by starry-shivam** [comment]
Facing same issue with `3.0.0-alpha06` on compose multiplatform.

**Comment by starry-shivam** [comment]
> Facing same issue with `3.0.0-alpha06` on compose multiplatform.


I'd like to add that setting a fixed height to either `AsyncImage` via a modifier, or if it's wrapped in another composable like `Box`, then adding a fixed height to that parent composable, resolves the lag issue. The problem seems to be that it does not perform lazy loading of items when there is no fixed height set, causing all of them to load at once. I discovered this when I attached a `DebugLogger()` to the Coil configuration, and it printed 960+ items at once after the screen rendered, even though my grid is only large enough to show 6 items at a time. After adding a fixed height, the logger printed 8 items initially and continued printing more as I scrolled further, indicating that it was indeed performing lazy loading this time. As a result, there were no lags or jitters.

**Comment by mainrs** [comment]
> > Facing same issue with `3.0.0-alpha06` on compose multiplatform.
> 
> I'd like to add that setting a fixed height to either `AsyncImage` via a modifier, or if it's wrapped in another composable like `Box`, then adding a fixed height to that parent composable, resolves the lag issue. The problem seems to be that it does not perform lazy loading of items when there is no fixed height set, causing all of them to load at once. I discovered this when I attached a `DebugLogger()` to the Coil configuration, and it printed 960+ items at once after the screen rendered, even though my grid is only large enough to show 6 items at a time. After adding a fixed height, the logger printed 8 items initially and continued printing more as I scrolled further, indicating that it was indeed performing lazy loading this time. As a result, there were no lags or jitters.

Thank you for sharing! This did work in one of my apps. There are some UI experiences where this is not possible, though. For example in staggered grid views where items can have different heights depending on their content. That's where I encountered the issue.

**Comment by starry-shivam** [comment]
> Thank you for sharing! This did work in one of my apps. There are some UI experiences where this is not possible, though. For example in staggered grid views where items can have different heights depending on their content. That's where I encountered the issue.

Yeah, this is just a workaround I found by trial and error. It cannot really cover all of the use cases. I really hope this bug gets fixed soon.

**Comment by colinrtwhite** [comment]
Hey folks if you're having performance issues I'd recommend trying the latest `3.0.0` alphas, as it should be faster. There's been work done in `AsyncImagePainter` to avoid Compose and Coroutines overhead. Some of this work required changing the public API slightly, which is why it hasn't been backported to `2.x` (though some slight improvements were backported to `2.7.0`). Here's the [upgrade guide](https://coil-kt.github.io/coil/upgrading_to_coil3/).

**Comment by hypersebi** [comment]
It got better after upgrading to 2.7.0, but still jittering. I've added a plain image (just for isolating coil) and I still got skipping frames on my 120hz pixel 7 pro. I suspect the Lazy Grid is becoming the issue from this point.

**Comment by JJSarrasin** [comment]
Same here, I develop on low-end TV-Box and if I have a grid of 4 columns. When I scroll down, the render is very laggy when it's not in cache and then it becomes usable, but still laggy. 
I try to set a fix size as suggested above, but no change.
I had better perfs with Picasso/views on same box, but it's a whole new world :)

**Comment by colinrtwhite** [comment]
Hi folks, I'm going to close this out as a lot of performance work went into `3.1.0` and I think Coil's in a decent spot. If you're still running into performance issues remember to test in release mode with R8 enabled (it makes a big difference). If you're still running into issues in release mode with R8 please open a new issue with reproducible test case.
